Rename Pattern to Bindings and share parsing logic with Assignment Target

MichaReiser opened the pull request:
Sorry, it seems this escalated quite a bit. I wasn't aware, that patterns are used in so many different places in our codebase

## Summary

#1805 separated assignment targets out from patterns. This PR now renames the "remaining" patterns to bindings. Bindings are different from assignment targets because they bind a new name in the current scope and assign a value to it rather than assigning a value to an existing name.  

This PR further restructures Patterns as specified in #1719. 

* Renames `Pattern` to `JsAnyBinding`
* Replaces `AssignPattern` with `JsBindingWithDefault`
* Deletes `PatternOrExpr`
* Replace `SinglePattern` with `JsIdentifierBinding`
* Replace `RestPattern` with `JsArrayRestBinding` and `JsObjectRestBinding`. Remove it from the `Pattern` union (only permitted inside of arrays/objects)
* Rename `ObjectPattern` to `JsObjectBinding`
* Replace `KeyValuePattern` with `JsShorthandPropertyBinding` and `JsPropertyBinding`
* Rename `ArrayPattern` to `JsArrayBinding`
* Remove `ExprPattern`
* Unify `SpreadElement` and `JsSpread`
* Deleted `InitializedProp` (not valid JS nor TS syntax???)

This PR extracts the common logic for parsing `Patterns` from `assignment_target` as traits into `pattern.rs` that I then used to parse bindings.

Part of #1725 

## Test Plan

Verified coverage, added new test cases.

All libs are now parsing without panicking. Performance is roughly the same

ematipico commented on crates/rome_formatter/src/ts/bindings/array_binding.rs:
```diff
@@ -0,0 +1,32 @@
+
+		Ok(format_elements![group_elements(format_elements![
+			l_bracket,
+			join_elements(space_token(), elements),
```
Should we use the same code we use for array elements?

```rust
			soft_indent(join_elements(soft_line_break_or_space(), elements)),
			if_group_breaks(token(",")),
```

MichaReiser replied:
Probably, but I just moved this logic from `ArrayPattern`. Will leave it, for now, to not change even more behaviour.

ematipico commented on crates/rslint_parser/src/parser/parsed_syntax.rs:
```diff
@@ -47,6 +47,16 @@ impl ParsedSyntax {
 		}
 	}
 
+	pub fn map<F>(self, mapper: F) -> ParsedSyntax
```
Can we document this new API, please? Also, I wouldn't call it only "map" because it's not a real map, it doesn't transform the input to "any" output.  But if this is the idiomatic pattern, I would describe it in the documentation

MichaReiser replied:
Sure! And you're right. Renamed it to `and_then` to match the method on [`Result`](https://doc.rust-lang.org/std/result/enum.Result.html#method.and_then)

ematipico commented on crates/rslint_parser/src/syntax/assignment_target.rs:
```diff
@@ -105,92 +106,27 @@ pub(crate) fn parse_simple_assignment_target(
 
+struct ArrayAssignmentTarget;
+
 // test array_assignment_target_rest
```
Are the comments still valid? I mean, should we move them next to the function that does the job?

MichaReiser replied:
The implementation is now shared between bindings and assignment targets. I still want to test both because they have slightly different behaviour. 

I moved it to `rest_pattern_kind` so that they're closer where they are used.

ematipico commented on crates/rslint_parser/src/syntax/binding.rs:
```diff
@@ -0,0 +1,273 @@
+use crate::parser::{expected_any, ToDiagnostic};
+#[allow(deprecated)]
+use crate::syntax::class::parse_equal_value_clause;
```
`TODO` to remove it?

MichaReiser replied:
Uh, not sure how that deprecated made it in there. Probably because I removed the `SingleTokenParseRecovery` use and it now moved to the `parse_equal_value_clause`.

ematipico commented on crates/rslint_parser/src/syntax/decl.rs:
```diff
@@ -145,7 +78,7 @@ pub(super) fn parse_parameters_list(
 				let m = p.start();
 				p.bump_any();
-				m.complete(p, JS_UNKNOWN_PATTERN);
+				m.complete(p, JS_UNKNOWN_BINDING);
```
`TODO`?

ematipico commented on crates/rslint_parser/src/syntax/expr.rs:
```diff
@@ -539,15 +541,28 @@ pub(super) fn any_reference_member(p: &mut Parser) -> Option<CompletedMarker> {
-		ident.change_kind(p, JS_REFERENCE_IDENTIFIER_MEMBER);
-		ident
-	})
+pub(super) fn is_at_reference_identifier_member(p: &Parser) -> bool {
```
Small doc?

MichaReiser replied:
What would you write in this documentation (same for `parse_reference_identifier_member`. I can only think of

> Tests if the parser is positioned at a reference identifier member

or 

> Parses a reference identifier member

which doesn't provide more information than the function names. Especially since `parse_*` is now a common pattern and `is_at_*` will be more common too.

ematipico replied:
I would describe/simplify the "reference identifier member" part. Also, I don't know what `is_keyword` does, which makes me think what is method doing.

MichaReiser replied:
Hmm... the `reference_identifier_member` is a name of an AST node. Yes, we should rename that node but we haven't found a better name yet. But that's why I can't simplify the method name.

`is_keyword` tests if that token is a `keyword`. For example, `for` is a valid member even tough it is a keyword.

MichaReiser replied:
This is the corresponding grammar: 

https://github.com/rome/tools/blob/914f0c2dd9087db42f5c24a3215f28a60d1aa8b1/xtask/js.ungram#L955-L961

And we have a name now... `JsName`, keep it simple :) Put we should to this in a separate PR

ematipico commented on crates/rslint_parser/src/syntax/expr.rs:
```diff
@@ -539,15 +541,28 @@ pub(super) fn any_reference_member(p: &mut Parser) -> Option<CompletedMarker> {
+	p.at(T![ident]) || p.cur().is_keyword()
+}
+
+pub(super) fn parse_reference_identifier_member(p: &mut Parser) -> ParsedSyntax {
```
Small doc?

ematipico commented on crates/rslint_parser/src/syntax/object.rs:
```diff
@@ -156,14 +156,6 @@ fn parse_object_member(p: &mut Parser) -> ParsedSyntax {
 				Present(m.complete(p, JS_METHOD_OBJECT_MEMBER))
 			} else if let Some(mut member_name) = member_name {
-				// test object_expr_assign_prop
-				// let b = { foo = 4, foo = bar }
```
No, it's not a valid syntax

ematipico commented on crates/rslint_parser/src/syntax/pattern.rs:
```diff
@@ -0,0 +1,306 @@
+///! Provides traits for parsing pattern like nodes
```
Should we expand on what kind of patterns these traits are meant to parse?

MichaReiser replied:
I can mention `array`, `object`, and `with_default` patterns but I guess that's already mentioned by the traits and will only become outdated if we change any of them.

ematipico commented on crates/rslint_parser/src/syntax/pattern.rs:
```diff
@@ -0,0 +1,306 @@
+	fn array_pattern_kind() -> SyntaxKind;
+	/// The kind of the rest pattern
+	fn rest_pattern_kind() -> SyntaxKind;
+	/// Creates a diagnostic saying that the parser expected an element at the passed in position
```
```suggestion
	/// Creates a diagnostic saying that the parser expected an element at an arbitrary position, passed as input
```

ematipico reviewed (commented):
It's a lot of code and it will take time for me to review. For now I leave some feedback and question:

- we should add new tests to the formatter for array binding. It seems we don't have them, other than `simple.js`;
- why did you use traits in this case? what's the endgame? The documentation doesn't explain that part and I can see you implemented the traits just once, which means that it could have been a `struct` or an `enum`;

MichaReiser commented:
> It's a lot of code and it will take time for me to review. For now I leave some feedback and question:

Thanks for taking the time! 

>     * we should add new tests to the formatter for array binding. It seems we don't have them, other than `simple.js`;

I'm not sure if that code is final. I just moved the files from the `pattern` to the `bindings` folder and renamed them. I don't know why git doesn't recognize the moves. I would leave it for now because this PR isn't touching the formatter code and we need to revisit all implementations anyway.

>     * why did you use traits in this case? what's the endgame? The documentation doesn't explain that part and I can see you implemented the traits just once, which means that it could have been a `struct` or an `enum`;

The traits are implemented twice: Once for `Bindings` and once for `AssignmentTarget`. For example, `ObjectAssignmentTarget` and `ArrayAssignmentTarget` both implement the `ParseObjectPattern` trait.

ematipico commented on crates/rslint_parser/src/syntax/binding.rs:
```diff
@@ -0,0 +1,285 @@
+	// let { ...c = "default" } = a;
+	// let { ...{a} } = b;
+	// let { ...rest, other_assignment } = a;
+	// let { ...rest, } = a;
```
Let's add a test for 
```js
let {  ...await } = a;
```
It should blow for incorrect use of the keyword `await`

MichaReiser replied:
This is actually valid, at least outside of async functions.

ematipico replied:
I just tried it with Babel, shift and espree. They all emit an error on incorrect use of `await`

ematipico commented on crates/rslint_parser/src/syntax/binding.rs:
```diff
@@ -0,0 +1,285 @@
+			if let Present(mut inner) = inner {
+				if inner.kind() != JS_IDENTIFIER_BINDING {
+					inner.change_kind(p, JS_UNKNOWN_BINDING);
+					p.error(p.err_builder("Expected identifier binding").primary(inner.range(p), "Object rest patterns must bind to an identifier, other patterns are not allowed."))
```
I wonder if we could suggest a way to fix the error based on what we have? Like, if we know we have brackets, to remove them; if it's a number to replace it with a correct name, etc.

MichaReiser replied:
That would be neat and should be possible. But let's wait fo the autofix API to do this.

ematipico commented on crates/rslint_parser/src/syntax/class.rs:
```diff
@@ -67,31 +73,34 @@ fn class(p: &mut Parser, kind: ClassKind) -> CompletedMarker {
+		match parse_identifier_binding(&mut *guard) {
+			Valid(Present(id)) => {
+				let text = guard.span_text(id.range(&*guard));
+				if guard.typescript() && DISALLOWED_TYPE_NAMES.contains(&text) {
+					let err = guard
+						.err_builder(&format!(
+							"`{}` cannot be used as a class name because it is already reserved as a type",
+							text
+						))
+						.primary(id.range(&*guard), "");
+
+					guard.error(err);
+				}
```
Do we really need typescript here? I mean, can we remove it for now?

MichaReiser replied:
That's copied from further below. I didn't change anything.

ematipico commented on crates/rslint_parser/src/syntax/decl.rs:
```diff
@@ -195,8 +129,8 @@ pub(super) fn parse_parameters_list(
 				p,
-				ParseRecovery::new(
-					JS_UNKNOWN_PATTERN,
+				&ParseRecovery::new(
```
I see that you changed it! Good :) 
I had to change too

MichaReiser replied:
hehe yeah, was annoying that it moved :)

ematipico commented on crates/rslint_parser/src/syntax/pattern.rs:
```diff
@@ -0,0 +1,306 @@
+
+		with_default
+			.parse_pattern(p)
+			.or_missing_with_error(p, |p, _| P::expected_pattern_error(p, rest_end..rest_end));
```
Is `rest_end...rest_end` right? 🤔

MichaReiser replied:
Yes, we want to point the error directly at the end of the rest token. 

```
[ ...    ]
    ^ // point right after the last dot and not..

[ ...     ]
         ^ // to the next token

ematipico commented on crates/rslint_parser/src/syntax/pattern.rs:
```diff
@@ -0,0 +1,306 @@
+					break;
+				}
+
+				match recover_result {
+					Err(RecoveryError::Eof) => break,
+					Err(RecoveryError::AlreadyRecovered) => {
+						guard.error(Self::expected_property_pattern_error(
+							guard,
+							guard.cur_tok().range,
+						));
+						break;
+					}
+					_ => {}
+				}
```
I don't understand this part. If `recover_result` is `Err`, it will never reach this `match` because it will hit the `break;`. Also, why marking the error *again*? I think `or_recover` already does that.

MichaReiser replied:
Nice catch! That makes no sense at all

ematipico commented on crates/rslint_parser/src/syntax/pattern.rs:
```diff
@@ -0,0 +1,306 @@
+			p.err_builder("rest element cannot have a default")
+				.primary(
+					default_start..p.cur_tok().range.start,
+					"Remove the default value here",
```
The default should also contain the `=`. The snapshot shows that the diagnostic has only the range of the default value.

I mean, if we remove only the default value, we still have invalid syntax because also the `=` should be gone.

MichaReiser replied:
Looks correct to me

```
error[SyntaxError]: rest element cannot have a default
  ┌─ rest_property_binding_err.js:2:12
  │
2 │ let { ...c = "default" } = a;
  │       ---- ^^^^^^^^^^^^ Remove the default value here
  │       │     
  │       Rest element
```

ematipico replied:
Sorry, I must have had some illusion 😆

ematipico commented on crates/rslint_parser/src/syntax/pattern.rs:
```diff
@@ -0,0 +1,306 @@
+	} else {
+		p.error(
+			p.err_builder("rest element must be the last element")
+				.primary(rest.range(p), "Move the rest element to the end"),
```
I would add `to the end of the pattern, right before the closing curly bracket`

MichaReiser commented on crates/rslint_parser/src/syntax/binding.rs:
```diff
@@ -0,0 +1,285 @@
+// let let = 5;
+// const let = 5;
+// let a, a;
+pub(crate) fn parse_identifier_binding(p: &mut Parser) -> ConditionalParsedSyntax {
```
Yes, that makes sense. I also need to revisit the `ConditionalParsedSyntax`. The way it currently works (where `Absent` values can be invalid too`) is super awkward to use.

MichaReiser replied:
I'm also still figuring this out. So far I thought it's a good idea to make the `or_invalid_to_unknown` conversion right inside of the parse function but it turned out that is dangerous. 

For example, the class or function id, a single arrow function parameter or an object rest pattern don't support `JsAnyBinding` and, therefore, returning a `JsUnknownBinding` is invalid in these cases. So it seems, all we can do in these cases is to bubble up the error until we come to a place where we can handle it. 

This is a bit annoying. An alternative would be to just add a diagnostic and keep the binding as is. This would certainly be easier but we then no longer have a strict rule when to use unknown:

* any syntax error (additional syntax) -> Content is inside of an `Unknown` node
* missing syntax -> Empty slot, handled in the AST facade

We need to get some more experience how to handle and we may come back to this change.

ematipico replied:
I bumped into this too while refactoring `class.rs`, where some function uses `or_invalid_to_unknown` under to hood but it is "hidden" and we don't know what's happening

MichaReiser commented:
Update on `await`. It's a valid identifier in scripts (outside of async) functions but not in modules. 

I added one more check to disallow `let` as an identifier in `strict mode` (see coverage improvements)
